contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol

```diff
@@ -0,0 +1,710 @@
+     * @param _asset Address of the asset
+     */
+    function supportsAsset(address _asset) public view override returns (bool) {
+        return _asset == ws;
+    }
+
+    /**
+     * @notice Collect accumulated SWPx (and other) rewards and send to the Harvester.
+     */
+    function collectRewardTokens()
+        external
+        override
+        onlyHarvester
+        nonReentrant
+    {
+        // Collect SWPx rewards from the gauge
+        IGauge(gauge).getReward();
+
+        _collectRewardTokens();
+    }
+
+    /***************************************
+        Internal SwapX Pool and Gauge Functions
+    ****************************************/
+
+    /**
+     * @dev Calculate the required amount of OS to mint based on the wS amount.
+     * This ensures the proportion of OS tokens being added to the pool matches the proportion of wS tokens.
+     * For example, if the added wS tokens is 10% of existing wS tokens in the pool,
+     * then the OS tokens being added should also be 10% of the OS tokens in the pool.
+     * @param _wsAmount Amount of Wrapped S (wS) to be added to the pool.
+     * @return osAmount Amount of OS to be minted and added to the pool.
+     */
+    function _calcTokensToMint(uint256 _wsAmount)
+        internal
+        view
+        returns (uint256 osAmount)
+    {
+        (uint256 wsReserves, uint256 osReserves, ) = IPair(pool).getReserves();
+        require(wsReserves > 0, "Empty pool");
+
+        // OS to add = (wS being added * OS in pool) / wS in pool
+        osAmount = (_wsAmount * osReserves) / wsReserves;
+    }
+
+    /**
+     * @dev Calculate how much pool LP tokens to burn to get the required amount of wS tokens back
+     * from the pool.
+     * @param _wsAmount Amount of Wrapped S (wS) to be removed from the pool.
+     * @return lpTokens Amount of SwapX pool LP tokens to burn.
+     */
+    function _calcTokensToBurn(uint256 _wsAmount)
+        internal
+        view
+        returns (uint256 lpTokens)
+    {
+        /* The rate between coins in the pool determines the rate at which pool returns
+         * tokens when doing balanced removal (remove_liquidity call). And by knowing how much wS
+         * we want we can determine how much of OS we receive by removing liquidity.
+         *
+         * Because we are doing balanced removal we should be making profit when removing liquidity in a
+         * pool tilted to either side.
+         *
+         * Important: A downside is that the Strategist / Governor needs to be
+         * cognizant of not removing too much liquidity. And while the proposal to remove liquidity
+         * is being voted on, the pool tilt might change so much that the proposal that has been valid while
+         * created is no longer valid.
+         */
+
+        (uint256 wsReserves, , ) = IPair(pool).getReserves();
+        require(wsReserves > 0, "Empty pool");
+
+        lpTokens = (_wsAmount * IPair(pool).totalSupply()) / wsReserves;
```

Comment:
To express the rounding up, and not round up you could also do: 
```
lpTokens = (_wsAmount * IPair(pool).totalSupply() + 1e18 - 1) / wsReserves;
```
Pro: doesn't round up when the division is perfect  / whole numbered 
Con: affects code readability negatively

Reply:
we want to round up to ensure we get enough LP tokens out of the gauge and then wS tokens out off the pool.

I originally had this code but testing showed cases where not enough LP tokens were used to get the required amount of wS tokens.
```
lpTokens =
            ((_wsAmount + 1) * IPair(pool).totalSupply()) /
            ws.balanceOf(pool);
```